### ISSUE TITLE
[v2] Large requests hang/become very slow

### ISSUE THREAD

**Issue by tzneal** [telepresenceio/telepresence#2092]
Large requests never finish. This may be a duplicate of #1953 , but I've included a reproducing test case.  This may be related to an interaction between kind & telepresence, I haven't tested on any other cluster.

### install a kind cluster
```
kind create cluster
```

### install the test service (https://gist.github.com/tzneal/44fa4d0e44e79ca9d711de76de16d97a)
```
kubectl apply -f telepresence-bug.yaml
```

### intercept
```
telepresence connect
telepresence intercept bug-demo --port 1234
```

### run the intercepted service (https://gist.github.com/tzneal/be5739e3d2e855d98efb894e36aa3ffa)
```
./serve.py 1234
```

### run curl and the request hangs
```
$ curl http://bug-demo
.... bunch of lines removed...
88365 ABCDEFGHIJKLMNOPQRSTUVWXYZ
88366 ABCDEFGHIJKLMNOPQRSTUVWXYZ
88367 ABCDEFGHIJKLMNOPQRSTUVWXYZ
*curl is now hung waiting on data*
```

**Comment by tzneal** [comment]
Something weird is going on, the transfer is hanging because the TCP window is full.

![wireshark-tcp](https://user-images.githubusercontent.com/64753/133804210-c905a0e0-c4a3-4fcc-a8f3-96db2d28159c.PNG)

**Comment by thallgren** [comment]
Thanks for providing this reproducer @tzneal . Can I assume that you've tested this with the 2.4.3 release (went out yesterday)?

**Comment by tzneal** [comment]
Yup, I just downloaded a fresh copy this morning to test with.
```
Client: v2.4.3 (api v3)
Root Daemon: v2.4.3 (api v3)
User Daemon: v2.4.3 (api v3)
```

**Comment by thallgren** [comment]
Great. This just moved to the very top of my list.

**Comment by tzneal** [comment]
Hopefully it's not just an oddity of my setup.  I've been looking into it as well and set rootDaemon to debug level logs.  You can look at the timestamps and see at around 10:29 it goes a little weird and then it's only keep alives.

```
2021-09-17 10:28:58.4189 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926669022, an 296576640, wz 2048, len 1460, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4582 debug   daemon/server-router/TUN reader : <- TUN tcp 10.96.0.0:52374 -> 10.96.221.181:80, sq 296576640, an 926670482, wz 14742, len 0, flags=ACK (from pkg/tun/tcp/handler.go:585)
2021-09-17 10:28:58.4583 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 0 (from pkg/tun/tcp/handler.go:722)
2021-09-17 10:28:58.4583 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 1 (from pkg/tun/tcp/handler.go:693)
2021-09-17 10:28:58.4584 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 2 (from pkg/tun/tcp/handler.go:693)
2021-09-17 10:28:58.4584 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 3 (from pkg/tun/tcp/handler.go:693)
2021-09-17 10:28:58.4584 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 4 (from pkg/tun/tcp/handler.go:693)
2021-09-17 10:28:58.4584 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 5 (from pkg/tun/tcp/handler.go:693)
2021-09-17 10:28:58.4585 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 6 (from pkg/tun/tcp/handler.go:693)
2021-09-17 10:28:58.4585 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 7 (from pkg/tun/tcp/handler.go:693)
2021-09-17 10:28:58.4584 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926670482, an 296576640, wz 2048, len 1460, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4585 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926671942, an 296576640, wz 2048, len 1460, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4585 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926673402, an 296576640, wz 2048, len 1460, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4585 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926674862, an 296576640, wz 2048, len 1460, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4585 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926676322, an 296576640, wz 2048, len 1460, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4586 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926677782, an 296576640, wz 2048, len 1460, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4586 debug   daemon/server-router/TUN writer : -> TUN tcp 10.96.221.181:80 -> 10.96.0.0:52374, sq 926679242, an 296576640, wz 2048, len 1296, flags=ACK,PSH (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:28:58.4886 debug   daemon/server-router/TUN reader : <- TUN tcp 10.96.0.0:52374 -> 10.96.221.181:80, sq 296576640, an 926680538, wz 15309, len 0, flags=ACK (from pkg/tun/tcp/handler.go:585)
2021-09-17 10:28:58.4886 debug   daemon/server-router/TUN reader :    CON tcp 10.96.0.0:52374 -> 10.96.221.181:80, Ack-queue size 0 (from pkg/tun/tcp/handler.go:722)
2021-09-17 10:28:59.0988 debug   daemon/server-router/TUN reader : -- POOL udp 10.96.0.0:59824 -> 10.96.0.10:53, count now is 4 (from pkg/connpool/pool.go:63)
2021-09-17 10:28:59.2235 debug   daemon/server-router/TUN reader : -- POOL udp 10.96.0.0:49137 -> 10.96.0.10:53, count now is 3 (from pkg/connpool/pool.go:63)
2021-09-17 10:29:00.3185 debug   daemon/server-router/TUN reader : <- TUN tcp 10.244.0.0:49156 -> 10.244.0.13:39293, sq 3073184753, an 825811141, wz 502, len 0, flags=ACK (from pkg/tun/tcp/handler.go:585)
2021-09-17 10:29:00.3186 debug   daemon/server-router/TUN reader : -> MGR tcp 10.244.0.0:49156 -> 10.244.0.13:39293, code KEEP_ALIVE (from pkg/tun/tcp/tmgr.go:128)
2021-09-17 10:29:00.3186 debug   daemon/server-router/TUN writer : -> TUN tcp 10.244.0.13:39293 -> 10.244.0.0:49156, sq 825811141, an 3073184754, wz 2048, len 0, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:29:15.4227 debug   daemon/server-router/TUN reader : <- TUN tcp 10.244.0.0:49156 -> 10.244.0.13:39293, sq 3073184753, an 825811141, wz 502, len 0, flags=ACK (from pkg/tun/tcp/handler.go:585)
2021-09-17 10:29:15.4228 debug   daemon/server-router/TUN reader : -> MGR tcp 10.244.0.0:49156 -> 10.244.0.13:39293, code KEEP_ALIVE (from pkg/tun/tcp/tmgr.go:128)
2021-09-17 10:29:15.4228 debug   daemon/server-router/TUN writer : -> TUN tcp 10.244.0.13:39293 -> 10.244.0.0:49156, sq 825811141, an 3073184754, wz 2048, len 0, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:29:30.5264 debug   daemon/server-router/TUN reader : <- TUN tcp 10.244.0.0:49156 -> 10.244.0.13:39293, sq 3073184753, an 825811141, wz 502, len 0, flags=ACK (from pkg/tun/tcp/handler.go:585)
2021-09-17 10:29:30.5264 debug   daemon/server-router/TUN reader : -> MGR tcp 10.244.0.0:49156 -> 10.244.0.13:39293, code KEEP_ALIVE (from pkg/tun/tcp/tmgr.go:128)
2021-09-17 10:29:30.5265 debug   daemon/server-router/TUN writer : -> TUN tcp 10.244.0.13:39293 -> 10.244.0.0:49156, sq 825811141, an 3073184754, wz 2048, len 0, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
2021-09-17 10:29:45.6302 debug   daemon/server-router/TUN reader : <- TUN tcp 10.244.0.0:49156 -> 10.244.0.13:39293, sq 3073184753, an 825811141, wz 502, len 0, flags=ACK (from pkg/tun/tcp/handler.go:585)
2021-09-17 10:29:45.6302 debug   daemon/server-router/TUN reader : -> MGR tcp 10.244.0.0:49156 -> 10.244.0.13:39293, code KEEP_ALIVE (from pkg/tun/tcp/tmgr.go:128)
2021-09-17 10:29:45.6303 debug   daemon/server-router/TUN writer : -> TUN tcp 10.244.0.13:39293 -> 10.244.0.0:49156, sq 825811141, an 3073184754, wz 2048, len 0, flags=ACK (from pkg/client/daemon/tunrouter.go:313)
```

**Comment by tzneal** [comment]
I saw some commits that looked semi-relevant to this issue, so I pulled a nightly to test but the issue is still there:

```
Client: v2.4.4-nightly-ff5f468e (api v3)
Root Daemon: v2.4.4-nightly-ff5f468e (api v3)
User Daemon: not running
```

**Comment by an-tex** [comment]
I have the same issue with requests 2MB (maxReceiveSize is set appropriately). It's pretty hard to pinpoint but falling back to telepresence-legacy it works straight out of the box :(

**Comment by thallgren** [comment]
This will be addressed in our coming 2.4.5 release. If you want, you can try out our release candidate: `https://app.getambassador.io/download/tel2/linux/amd64/2.4.5-rc.1/telepresence` (swap linux for darwin on a macOS). Both the telepresence client and the traffic-manager must be reinstalled for the fix to kick in.

**Comment by an-tex** [comment]
I gave the 2.4.5-rc.1 a quick spin with a local k3d cluster and it seems to work! Had quite a few error logs in the traffic manager along the lines of:
`2021-10-15 14:15:32.5099 error   httpd/conn=127.0.0.1:8081 :    CONN tcp 10.43.0.0:54110 -> 10.43.227.61:80 stream to-conn loop ended because context canceled` but it did't seem to affect the connections. Thanks @thallgren :)

**Comment by thallgren** [comment]
@an-tex thanks for reporting that. I'll investigate why those lines are printed.

**Comment by tzneal** [comment]
@thallgren  I did some testing this morning and this looks like it's resolved in the 2.4.5 release. Thanks!

**Comment by khussey** [comment]
Closing now that the 2.4.5 release is available.
